File: cabot_influxdb_client/script/pose_log_client.py
#!/usr/bin/env python3

import rclpy
from rclpy.node import Node
from datetime import datetime
from cabot_msgs.msg import PoseLog
from influxdb_client import InfluxDBClient, Point, WriteOptions
from influxdb_client.client.write_api import SYNCHRONOUS


USERNAME = "cabot"
PASSWORD = "cabot-influxdb"
TOKEN = "a54a87f7-73a0-4534-9741-ad7ff4e7d111"
ORGANIZATION_NAME = "cabot"
BUCKET_NAME = "grafana-test"
host = "localhost"
port = 8086

influxdb_url = f"http://{host}:{port}"
token = TOKEN
org = ORGANIZATION_NAME
bucket = BUCKET_NAME

client = InfluxDBClient(url=influxdb_url, token=token, org=org)
write_api = client.write_api(write_options=SYNCHRONOUS)


class PoseLogListener(Node):
    def __init__(self):
        super().__init__('pose_log_listener')
        self.subscription = self.create_subscription(PoseLog, '/cabot/pose_log', self.pose_log_callback, 10)
        self.client = InfluxDBClient(url='http://localhost:8086', token=token, org=org)
        self.write_api = self.client.write_api(write_options=SYNCHRONOUS)

    def pose_log_callback(self, msg):
        point = Point("pose_data").field("lat", msg.lat).field("lng", msg.lng).field("floor", msg.floor).time(datetime.utcnow())
        self.write_api.write(bucket=bucket, org=org, record=point)

    def __del__(self):
        self.client.close()

def main(args=None):
    rclpy.init(args=args)
    pose_log_listener = PoseLogListener()
    rclpy.spin(pose_log_listener)
    pose_log_listener.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main()

File: cabot_influxdb_client/src/pose_log_client.cpp
#include "pose_log_client.hpp"

const std::string USERNAME = "cabot";
const std::string PASSWORD = "cabot-influxdb";
const std::string TOKEN = "";
const std::string ORGANIZATION_NAME = "cabot";
const std::string BUCKET_NAME = "grafana-test";
const std::string INFLUXDB_URL = "http://localhost:8086";


PoseLogListener::PoseLogListener() : Node("pose_log_listener"){
  subscription_ = create_subscription<cabot_msgs::msg::PoseLog>(
      "/cabot/pose_log", 10, std::bind(&PoseLogListener::pose_log_callback, this, std::placeholders::_1));
  RCLCPP_INFO(this->get_logger(), "Node is ready");
}

void PoseLogListener::pose_log_callback(const cabot_msgs::msg::PoseLog::SharedPtr msg){
  influxdb::Point point("pose_data");
  point.addField("lat", msg->lat);
  point.addField("lng", msg->lng);
  point.addField("floor", msg->floor);
  point.setTimestamp(std::chrono::system_clock::now());

  try{
    std::unique_ptr<influxdb::InfluxDB> idb(influxdb::InfluxDBFactory::Get(INFLUXDB_URL));
    idb->write(std::move(point));
  }catch(const std::exception& e){
    RCLCPP_ERROR(this->get_logger(), "Error writing to InfluxDB: %s", e.what());
  }
}

int main(int argc, char* argv[]){
  rclcpp::init(argc, argv);
  std::unique_ptr<PoseLogListener> node_ptr(std::make_unique<PoseLogListener>());
  std::shared_ptr<PoseLogListener> node(std::move(node_ptr));
  rclcpp::spin(node);
  rclcpp::shutdown();
  return 0;
}

